radicle-node/src/wire.rs

```diff
@@ -423,51 +425,68 @@ impl Decode for node::Features {
 }
 
 #[derive(Debug)]
-pub struct Wire<R, S, T, G> {
-    inboxes: HashMap<IpAddr, Decoder>,
-    inner: service::Service<R, S, T, G>,
+pub struct Wire<R, S, W, G, T: Transcode> {
+    inboxes: HashMap<net::SocketAddr, Parser>,
+    inner: service::Service<R, S, W, G>,
+    transcoder: T,
+    handshake_queue: VecDeque<(net::SocketAddr, Vec<u8>)>,
 }
 
-impl<R, S, T, G> Wire<R, S, T, G> {
-    pub fn new(inner: service::Service<R, S, T, G>) -> Self {
+impl<R, S, W, G, T: Transcode> Wire<R, S, W, G, T> {
+    pub fn new(inner: service::Service<R, S, W, G>, transcoder: T) -> Self {
         Self {
             inboxes: HashMap::new(),
             inner,
+            transcoder,
+            handshake_queue: Default::default(),
         }
     }
 }
 
-impl<R, S, T, G> Wire<R, S, T, G>
+impl<R, S, W, G, T> Wire<R, S, W, G, T>
 where
     R: routing::Store,
     S: address::Store,
-    T: WriteStorage + 'static,
+    W: WriteStorage + 'static,
     G: Signer,
+    T: Transcode,
 {
-    pub fn connected(
-        &mut self,
-        addr: std::net::SocketAddr,
-        local_addr: &std::net::SocketAddr,
-        link: Link,
-    ) {
-        self.inboxes.insert(addr.ip(), Decoder::new(256));
+    pub fn connected(&mut self, addr: net::SocketAddr, local_addr: &net::SocketAddr, link: Link) {
+        self.inboxes.insert(addr, Parser::new(256));
         self.inner.connected(addr, local_addr, link)
     }
 
     pub fn disconnected(
         &mut self,
-        addr: &std::net::SocketAddr,
+        addr: &net::SocketAddr,
         reason: nakamoto::DisconnectReason<service::DisconnectReason>,
     ) {
-        self.inboxes.remove(&addr.ip());
+        self.inboxes.remove(addr);
         self.inner.disconnected(addr, &reason)
     }
 
-    pub fn received_bytes(&mut self, addr: &std::net::SocketAddr, bytes: &[u8]) {
-        let peer_ip = addr.ip();
+    pub fn received_bytes(&mut self, addr: &net::SocketAddr, raw_bytes: &[u8]) {
+        let bytes = match self.transcoder.input(raw_bytes) {
+            // The received bytes were consumed by transcoder updating its inner
+            // state (for instance, this happens during handshake).
+            Ok(DecodedData::Local(bytes)) | Ok(DecodedData::Remote(bytes)) if bytes.is_empty() => {
+                return
+            }
+            Ok(DecodedData::Local(bytes)) => bytes,
+            Ok(DecodedData::Remote(output)) => {
+                log::trace!("performing handshake, sending to the remote peer more data");
+                self.handshake_queue.push_back((*addr, output));
+                return;
+            }
+            Err(err) => {
+                // TODO: Disconnect peer.
+                log::error!("invalid transcoder input. Details: {}", err);
+                return;
+            }
+        };
```

Comment:
Wouldn't it be simpler if we got rid of `DecodedData`, and in the case of `DecodedData::Remote`, we just handle it internally in the transcoder? I'm a little confused on when/how the transcoder switches between local and remote, as for example in the case of a handshake, it would seem that a transition between `Remote -> Local` would happen once and then all further outputs would be `Local`?

Reply:
We can't handle it in the transcoder since transcoder can't do a write - or doesn't maintains the queue of the events generated for the reactor. Here we are restricted by the specific reactor design from the nakamoto crate

Reply:
Sure, but we're keeping this Remote/Local distinction even after the handshake is over, instead of moving to a different transcoder. My suggestion would be to write it such that we start with eg. a `HandshakeTranscoder` for each peer, and then move on to a `WireTranscoder`, instead of having a single type that handles both. So it would be structured like:

```rust
struct State {
   negotiating: HashMap<Address, HandshakeTranscoder>, // Peers going through handshake
   negotiated: HashMap<Address, WireTranscoder>, // Peers after handshake
}
```